addon/components/listbox.js

```diff
@@ -0,0 +1,316 @@
+      }
+    } else if (event.key === 'ArrowLeft') {
+      if (this.isOpen) {
+        this.setPreviousOptionActive();
+      }
+    } else if (event.key === 'Home' || event.key === 'PageUp') {
+      this.setFirstOptionActive();
+    } else if (event.key === 'End' || event.key === 'PageDown') {
+      this.setLastOptionActive();
+    } else if (event.key === 'Escape') {
+      this.close();
+    }
+  }
+
+  @action
+  handleKeyPress(event) {
+    if (
+      event.key === 'Enter' ||
+      ((event.key === 'Space' || event.key === ' ') && this.search === '')
+    ) {
+      this.activateBehaviour = ACTIVATE_FIRST;
+      if (this.isOpen) {
+        this.setSelectedOption(undefined, event);
+        this.close();
+      } else {
+        this.open();
+      }
+    } else if (event.key.length === 1) {
+      this.addSearchCharacter(event.key);
+    }
+  }
+  @action
+  handleLabelClick(e) {
+    e.preventDefault();
+    if (e.ctrlKey) return;
+    this.buttonElement.focus();
+  }
+
+  @action
+  registerButtonElement(buttonElement) {
+    this.buttonElement = buttonElement;
+  }
+
+  @action
+  registerDocumentClickListener() {
+    this._handleDocumentClick = this.handleDocumentClick.bind(this);
+
+    document.addEventListener('click', this._handleDocumentClick);
+  }
+
+  @action
+  registerLabelElement(labelElement) {
+    this.labelElement = labelElement;
+  }
+
+  @action
+  registerOptionElement(optionValue, optionElement) {
+    this.optionElements.push(optionElement);
+    this.optionValues.push(optionValue);
+
+    if (this.args.value) {
+      if (this.args.value === optionValue) {
+        this.selectedOptionIndex = this.activeOptionIndex =
+          this.optionElements.length - 1;
+      }
+    }
+
+    if (!this.selectedOptionIndex) {
+      switch (this.activateBehaviour) {
+        case ACTIVATE_FIRST:
+          this.setFirstOptionActive();
+          break;
+        case ACTIVATE_LAST:
+          this.setLastOptionActive();
+          break;
+      }
+    }
+  }
+
+  @action
+  registerOptionsElement(optionsElement) {
+    this.optionsElement = optionsElement;
+    this.optionsElement.focus();
+  }
+
+  @action
+  setActiveOption(optionComponent) {
+    this.optionElements.find((o, i) => {
+      if (
+        o.id === optionComponent.guid + '-option' &&
+        !o.hasAttribute('disabled')
+      ) {
+        this.activeOptionIndex = i;
+        document.querySelector('#' + optionComponent.guid + '-option').focus();
+      }
+    });
+  }
+
+  @action
+  setSelectedOption(optionComponent, e) {
+    let optionId, optionValue, found;
+
+    if (optionComponent) {
+      optionId = optionComponent.guid + '-option';
+      optionValue = optionComponent.args.value;
+    } else if (this.activeOptionIndex !== undefined) {
+      optionId = this.optionElements[this.activeOptionIndex].id;
+      optionValue = this.optionValues[this.activeOptionIndex];
+    } else {
+      return;
+    }
+
+    this.optionElements.find((o, i) => {
```

Comment:
I think there are a few opportunities for refactoring with this functionality here

- We're using `.find`, which returns a specific member of `optionElements`, but don't actually ever _use_ the value that it's returning
- We're keeping track of whether the element has been located with a separate value, where that shouldn't be necessary; "finding" an element is already what `.find` does
- We're performing side-effects within `.find`, which isn't necessarily _wrong_ but is an unusual way of using the function

I would use `.find` to locate an element and then move all of the usage of the located element outside of the `.find` callback instead. We can simplify the code and remove the `found` state variable that way!

Reply:
https://github.com/GavinJoyce/ember-headlessui/pull/76/commits/9779e56ec174ea9f18cf492f122235bf0da05df2 refactors this method to remove the `find` call altogether.

In order to improve performance in large lists, the Listbox component now maintains an index on each option component and element. This avoids use of `find` altogether, as the option can now be referenced by its index in the list.